textdescriptives/components/quality.py

```diff
@@ -4,50 +4,121 @@
 from typing import Callable, Dict, List, Optional, Tuple, Union
 
 import numpy as np
+from pydantic import BaseModel, Field
 from spacy.language import Language
 from spacy.tokens import Doc, Span
 
+Interval = Tuple[Optional[float], Optional[float]]
 
-DEFAULT_QUALITY_THRESHOLDS = {
-    "n_stop_words": (2, None),
-    "alpha_ratio": (0.8, None),
-    "mean_word_length": (3, 10),
-    "doc_length": (10, 100_000),
-    "symbol_#_2_word_ratio": (None, 0.1),
-    "proportion_ellipsis": (None, 0.3),
-    "proportion_bullet_points": (None, 0.8),
-    "duplicate_line_chr_fraction": (None, 0.2),
-    "duplicate_paragraph_chr_fraction": (None, 0.2),
-    "duplicate_5-gram_chr_fraction": (None, 0.15),
-    "duplicate_6-gram_chr_fraction": (None, 0.14),
-    "duplicate_7-gram_chr_fraction": (None, 0.13),
-    "duplicate_8-gram_chr_fraction": (None, 0.12),
-    "duplicate_9-gram_chr_fraction": (None, 0.11),
-    "duplicate_10-gram_chr_fraction": (None, 0.1),
-    "top_2-gram_chr_fraction": (None, 0.20),
-    "top_3-gram_chr_fraction": (None, 0.18),
-    "top_4-gram_chr_fraction": (None, 0.16),
-    "contains_lorem ipsum": False,
-}
-
-
-def n_stop_words(span: Span) -> int:
+
+class QualityThresholds(BaseModel):
+    """Thresholds for quality metrics."""
+
+    n_stop_words: Interval = Field(
+        (2, None),
+        description="A Range for the number of stop words. Default: (2, None), i.e. "
+        + "at least 2 stop words, but no upper limit.",
+    )
+    alpha_ratio: Interval = Field(
+        (0.8, None),
+        description="A Range for the alpha ratio. Default: (0.8, None), i.e. at "
+        + r"least 80% of tokens contain at least one alphabetic character, but no "
+        + "upper limit.",
+    )
+    mean_word_length: Interval = Field(
+        (3, 10),
+        description="A Range for the mean word length. Default: (3, 10), i.e. between"
+        + " 3 and 10 characters.",
+    )
+    doc_length: Interval = Field(
+        (10, 100_000),
+        description="A Range for the document length. Default: (10, 100_000), i.e."
+        + " between 10 and 100_000 characters.",
+    )
+    symbol_to_word_ratio: Dict[str, Interval] = Field(
+        {"#": (None, 0.1)},
+        description="A dict of symbols and the allowed range for the "
+        + r"symbol-to-word-ratio. Default: {'#': (None, 0.1)} i.e. no lower limit, "
+        + r"but at most 10% of words start with a hashtag. Values not in the dict "
```

Comment:
Does this check how many words start with the symbol or how many symbols there are compared to words?

Reply:
It calculates the ratio between symbol occurrence and word occurrence and checks whether it is below a threshold  0.1. (so if we have 100 words the symbol should appear no more than 10 times).

Reply:
can you update the description to reflect that? (copy paste the above is fine)